src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/EnumConverterTests.cs

```diff
@@ -554,5 +555,85 @@ public enum SampleEnumSByte : sbyte
             F = 1 << 5,
             G = 1 << 6,
         }
+
+        [Fact]
+        public static void Honor_EnumNamingPolicy_On_Deserialization()
+        {
+            JsonSerializerOptions options = new()
+            {
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy: new SimpleSnakeCasePolicy() )
+                }
+            };
+
+            BindingFlags bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""non_public""", options);
+            Assert.Equal(BindingFlags.NonPublic, bindingFlags);
+
+            // Flags supported without naming policy.
+            bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""NonPublic, Public""", options);
+            Assert.Equal(BindingFlags.NonPublic | BindingFlags.Public, bindingFlags);
+
+            // Flags supported with naming policy.
+            bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""static, public""", options);
+            Assert.Equal(BindingFlags.Static | BindingFlags.Public, bindingFlags);
+
+            // Null not supported.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<BindingFlags>("null", options));
+
+            // Null supported for nullable enum.
+            Assert.Null(JsonSerializer.Deserialize<BindingFlags?>("null", options));
+        }
+
+        [Fact]
+        public static void EnumDictionaryKeyDeserialization()
+        {
+            JsonNamingPolicy snakeCasePolicy = new SimpleSnakeCasePolicy();
+            JsonSerializerOptions options = new()
+            {
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy: snakeCasePolicy)
+                },
+                DictionaryKeyPolicy = snakeCasePolicy
+            };
+
+            // Baseline.
+            var dict = JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""NonPublic, Public"": 1}", options);
+            Assert.Equal(1, dict[BindingFlags.NonPublic | BindingFlags.Public]);
+
+            // DictionaryKeyPolicy not honored for dict key deserialization.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""NonPublic0, Public0"": 1}", options));
+
+            // EnumConverter naming policy not honored.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""non_public, static"": 0, ""NonPublic, Public"": 1}", options));
+        }
+
+        [Fact]
+        public static void EnumDictionaryKeySerialization()
```

Comment:
add a test where enum converter (set through options) and properties have different naming policy and make sure both scenarios serialize/deserialize correctly

Reply:
Sounds good. It's essentially a test that the converter precedence rules hold (property-level converter wins over type-level), but the redundancy is good.